### ISSUE TITLE
Liquibase 4.23.x compatiblity

### ISSUE THREAD

**Issue by dinerotah** [liquibase/liquibase-groovy-dsl#55]
Hi,
When using liquibase 4.23.2 I have the following error:

Caused by: groovy.lang.MissingMethodException: No signature of method: liquibase.changelog.DatabaseChangeLog.include() is applicable for argument types: (String, Boolean, liquibase.integration.spring.SpringResourceAccessor...) values: [liquibase/patches/IRIS-27434-API-Possibility-to-configure-list-of-Roles-of-a-Project-per-Instance.groovy, ...]
Possible solutions: include(java.lang.String, boolean, boolean, liquibase.resource.ResourceAccessor, liquibase.ContextExpression, liquibase.LabelExpression, java.lang.Boolean, boolean), include(java.lang.String, boolean, boolean, liquibase.resource.ResourceAccessor, liquibase.ContextExpression, liquibase.Labels, java.lang.Boolean, boolean), include(java.lang.String, boolean, boolean, liquibase.resource.ResourceAccessor, liquibase.ContextExpression, liquibase.Labels, java.lang.Boolean, liquibase.changelog.DatabaseChangeLog$OnUnknownFileFormat), include(java.lang.String, boolean, boolean, liquibase.resource.ResourceAccessor, liquibase.ContextExpression, liquibase.Labels, java.lang.Boolean, liquibase.changelog.DatabaseChangeLog$OnUnknownFileFormat, liquibase.changelog.DatabaseChangeLog$ModifyChangeSets)
	at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.unwrap(ScriptBytecodeAdapter.java:72)
	at org.codehaus.groovy.runtime.callsite.PojoMetaClassSite.call(PojoMetaClassSite.java:48)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.liquibase.groovy.delegate.DatabaseChangeLogDelegate.include(DatabaseChangeLogDelegate.groovy:197)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:328)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:342)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1006)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at liquibase.Script1$_run_closure1.doCall(Script1.groovy:4)
	at liquibase.Script1$_run_closure1.doCall(Script1.groovy)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:328)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:279)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1006)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:39)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:130)
	at liquibase.parser.ext.GroovyLiquibaseChangeLogParser.processDatabaseChangeLogRootElement(GroovyLiquibaseChangeLogParser.groovy:130)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:328)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:342)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:63)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:49)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:203)
	at liquibase.parser.ext.GroovyLiquibaseChangeLogParser$_getChangeLogMethodMissing_closure3.doCall(GroovyLiquibaseChangeLogParser.groovy:88)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaMethod.invoke(ClosureMetaMethod.java:88)
	at groovy.lang.MetaClassImpl.invokeMissingMethod(MetaClassImpl.java:923)
	at groovy.lang.MetaClassImpl.invokePropertyOrMissing(MetaClassImpl.java:1412)
	at groovy.lang.MetaClassImpl.doInvokeMethod(MetaClassImpl.java:1334)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1087)
	at groovy.lang.ExpandoMetaClass.invokeMethod(ExpandoMetaClass.java:1142)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1006)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at liquibase.Script1.run(Script1.groovy:3)
	at liquibase.Script1$run.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:130)
	at liquibase.parser.ext.GroovyLiquibaseChangeLogParser.parse(GroovyLiquibaseChangeLogParser.groovy:59)

Could please check ??

**Comment by aminabromand** [comment]
same here

**Comment by dinerotah** [comment]
@stevesaliman  @recampbell  Any plan to make it compatible ?

**Comment by stevesaliman** [comment]
There seems to be a lot of changes to the Liquibase core from version to version starting with 4.17, and it is a challenge to keep up with them all.  I've been talking to the Liquibase team about ways to work around this, but nothing as of yet.

**Comment by dinerotah** [comment]
@stevesaliman I have a branch with fixes but I could not push 
`fatal: unable to access 'https://github.com/liquibase/liquibase-groovy-dsl.git/': The requested URL returned error: 403
`

**Comment by dinerotah** [comment]
@stevesaliman here is the patch:
[Liquibase_4_23_x_compatiblity_#55.txt](https://github.com/liquibase/liquibase-groovy-dsl/files/13298121/Liquibase_4_23_x_compatiblity_.55.txt)
Just change the extension to .patch

**Comment by stevesaliman** [comment]
The larger issue is backwards compatibility.  If we fix it for one version, we break it for another.  I'm trying to figure out how to figure out how to force users to upgrade to a new version of Liquibase when they want to use a new version of the DSL

**Comment by dinerotah** [comment]
@stevesaliman so just increase major version 4.0.0 which would be compatible with liquibase 4.17.+ and 3.x.x is for version <= 4.16.x

**Comment by dinerotah** [comment]
Like for 3.0.0:
[September 5, 2020](https://github.com/liquibase/liquibase-groovy-dsl#september-5-2020)
Version 3.0.0 of the Liquibase Groovy DSL is now available with support for Liquibase 4.0.0.

As you might expect for a major release, this means some breaking changes. There are two breaking changes with this release.

Version 3.0.0 of the DSL no longer supports the 3.x releases of Liquibase. If you need to use an older version of Liquibase, you'll need an older version of the DSL.

Liquibase 4.0.0 no longer supports using absolute filenemes, so the DSL doesn't either. This change only affects changelogs that were using the include and includeAll elements with absolute paths.

**Comment by dinerotah** [comment]
Hi @stevesaliman any update ? now it is blocking us to upgrade to springboot 3.2.0 as downgrading liquibase does not help anymore. 
java.lang.NoClassDefFoundError: liquibase/UpdateSummaryEnum and this class was added in liquibase 4.23.1

**Comment by stevesaliman** [comment]
Unfortunately there is no update for this issue.  I need to find a way of implementing this that doesn't break with every Liquibase release, and I'll likely need some support from the Liquibase team to get this stable.

**Comment by dinerotah** [comment]
@stevesaliman Releasing a major version for each liquibase breaking changes is the way to go. The liquibase team could help you by notifying you. 
We start thinking to remove this dependency and to switch to another format :-(

**Comment by WilliamHinschink** [comment]
Hi! I opened a [PR](https://github.com/liquibase/liquibase-groovy-dsl/pull/58) updating liquibase to version 4.27
Solving this problem **No signature of method**, that was caused by method signatures no longer existing in new versions of liquibase

**Comment by multipristr** [comment]
> The larger issue is backwards compatibility. If we fix it for one version, we break it for another. I'm trying to figure out how to figure out how to force users to upgrade to a new version of Liquibase when they want to use a new version of the DSL

If the DSL shouldn't work with older liquibase-core, can't you add [dependency versions constrains](https://docs.gradle.org/current/userguide/rich_versions.html) to [liquibase.gradle](https://github.com/liquibase/liquibase-groovy-dsl/blob/0d27379258f188948d688fb94361c15e599885f5/liquibase.gradle#L8)?
For example the following shouldn't compile if liquibase-core version is lower than 4.21.0:
```groovy
dependencies {
  runtime('org.liquibase:liquibase-core') {
            version {
                strictly '[4.21.0,)'
                prefer '4.27.1'
                because 'This version of liquibase-groovy-dsl requires liquibase-core 4.21.0 or higher'
            }
        }
}
```
@stevesaliman

**Comment by stevesaliman** [comment]
Version 4.0.0 has just been released, with support for Liquibase 4.24+, and de-supporting anything older.  This should take care of the problem in this Issue.
